sys/riscv/riscv/swtch.S

```diff
@@ -401,9 +597,31 @@ ENTRY(cpu_switch)
 	/*
 	 * Restore the saved context.
 	 */
+#ifdef __CHERI_PURE_CAPABILITY__
+	clc	c13, TD_PCB(ca1)
+#else
 	ld	x13, TD_PCB(a1)
+#endif
 
-#if __has_feature(capabilities)
+#ifdef __CHERI_PURE_CAPABILITY__
+	/* Restore the registers */
+	clc	cra, (PCB_RA)(c13)
+	clc	csp, (PCB_SP)(c13)
+
+	/* s[0-11] */
+	clc	cs0, (PCB_S + 0 * 16)(c13)
+	clc	cs1, (PCB_S + 1 * 16)(c13)
+	clc	cs2, (PCB_S + 2 * 16)(c13)
+	clc	cs3, (PCB_S + 3 * 16)(c13)
+	clc	cs4, (PCB_S + 4 * 16)(c13)
+	clc	cs5, (PCB_S + 5 * 16)(c13)
+	clc	cs6, (PCB_S + 6 * 16)(c13)
+	clc	cs7, (PCB_S + 7 * 16)(c13)
+	clc	cs8, (PCB_S + 8 * 16)(c13)
+	clc	cs9, (PCB_S + 9 * 16)(c13)
+	clc	cs10, (PCB_S + 10 * 16)(c13)
+	clc	cs11, (PCB_S + 11 * 16)(c13)
+#elif __has_feature(capabilities)
 	/* Restore the registers */
 	lc	cra, (PCB_RA)(x13)
```

Comment:
Hm, why did we make cpu_switch save/restore capabilities for a hybrid kernel? That shouldn't be necessary given the calling convention?

Reply:
Uhm probably we still want that to prevent leaking from one side to the other. @bsdjhb may have another opinion on this.

Reply:
This is the context switch code not the trap handler code, you only need to save/restore things that the kernel caller won't already have saved (and for hybrid it will have to save any capabilities that are live due to the calling convention). Restoring integers instead will stop capability leakage just as much, but that does raise an important question about whether we should be clearing the tags (or entire registers) for caller-saved registers, as currently those leak regardless of whether you're a hybrid or pure-capability kernel. Probably fine without any compartmentalisation and if you trust the compiler, and compartmentalisation will clear things as necessary at compartment boundaries anyway.

Reply:
What we do for hybrid already isn't relevant for the purecap kernel though, just something I noticed that we should probably fix in the hybrid kernel.

Reply:
Yeah, I was thinking as if kernel threads were running in different protection domains but it is definitely not the case. I removed the capability load/stores to the pcb for the hybrid kernel as a separate commit. I think for now we don't care about clearing tags as the next kthread can just access the root capability anyway, but we will with kernel compartments eventually.

Reply:
Probably because if we were bothering to save the register anyway, we might as well save the full ones.  This will break kgdb btw as it "knows" the layout of 'struct pcb' for plain RISC-V vs hybrid vs purecap.  Saving the full caps in hybrid does mean that you can at last see the full cap register contents for the kernel threads in kgdb.  pcb_ra at least will be a full cap even if the sX registers should all in theory be null-derived.

Reply:
(I can of course fix kgdb, but it assumes this layout currently)

Reply:
Let's keep the hybrid version for now, we can deal with removing that unnecessary extra case separately. pcb_ra being a full cap isn't helpful, you'll have just called cpu_switch with a plain JALR so ra should be an integer too.